Improve stereotypes support

Ladicek opened the pull request:
This commit adds a better API for registering custom stereotypes, similar
to existing APIs: `StereotypeRegistrar` and `StereotypeRegistrarBuildItem`.
The existing `AdditionalStereotypeBuildItem` is very confusing to use, so
it is deprecated and all its usages are replaced with the new API.

Also, this commit allows defining stereotypes for synthetic beans. Scope,
name, alternative status and priority are all applied to the synthetic bean,
but interceptor bindings are not. We don't apply interceptors to synthetic
beans in general, this case is not an exception.

Ladicek commented:
Creating as draft to let CI run in my fork, and to discuss the TODO items with @mkouba.

Ladicek commented:
Rebased. @mkouba could you please take a look? Thank you!

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/StereotypeRegistrar.java:
```diff
@@ -0,0 +1,17 @@
+
+    /**
+     * Returns a set of annotation types (their names) that should be treated as stereotypes.
+     * To modify (meta-)annotations on these annotations, use {@link AnnotationsTransformer}.
```
This is pretty smart (and I can see that it's even tested ;-)

Ladicek replied:
This is a very nice thing about ArC's extensions: annotation transformers apply to annotations. Portable Extensions don't allow that, and so to allow portability, neither do Build Compatible Extensions :-/

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -775,18 +778,23 @@ private Map<DotName, StereotypeInfo> findStereotypes(IndexView index, Map<DotNam
                         isNamed = true;
                     } else if (DotNames.PRIORITY.equals(annotation.name())) {
                         alternativePriority = annotation.value().asInt();
+                    } else if (DotNames.ARC_PRIORITY.equals(annotation.name()) && alternativePriority == null) {
```
Good catch. Do we have a test for it?

Ladicek replied:
Good question, I don't think so. I'll add one.

Ladicek replied:
Actually one of the tests I added in this PR uses ArC's `@AlternativePriority`, so that is actually covered. There's indeed no test for ArC's `@Priority`.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfigurator.java:
```diff
@@ -43,6 +43,31 @@ public void done() {
+
+            Integer priority = this.priority;
+            if (priority == null) {
+                // TODO use `Beans.initAlternativePriority` instead? IMHO no for synthetic beans
```
I think that `Beans.initAlternativePriority()` is purely about setting the priority to `Integer.MAX_VALUE` for alternatives selected via the `quarkus.arc.selected-alternatives` config property. Which probably does not make sense for synthetic beans...

Ladicek replied:
Good point, I didn't look deep enough when writing this. Will remove the comment.

mkouba commented on extensions/smallrye-health/deployment/src/main/java/io/quarkus/smallrye/health/deployment/SmallRyeHealthProcessor.java:
```diff
@@ -363,7 +363,10 @@ AnnotationsTransformerBuildItem annotationTransformer(BeanArchiveIndexBuildItem
             }
         }
-        List<DotName> healthAnnotations = new ArrayList<>(5);
+        // TODO additional stereotypes? see AutoProducerMethodsProcessor.annotationTransformer
```
Could you be more specific here?

Ladicek replied:
I (almost accidentally) stumbled upon this code in the SmallRye Health extension which automatically adds `@Singleton` to classes or methods that don't declare a scope but do declare a MicroProfile Health annotation. This was added by certain @mkouba in https://github.com/quarkusio/quarkus/pull/9749 :-)

To figure out if the class or method in question declares a scope or not, it uses `CustomScopeAnnotationsBuildItem.isScopeIn()` with 1. the set of annotations on the class or method, 2. the set of (meta-)annotations on anything annotated `@Stereotype`. This ignores custom stereotypes, which is what the TODO I added here highlights.

Thinking about it again, I have no idea why `SmallRyeHealthProcessor.annotationTransformer` even exists. `SmallRyeHealthProcessor.build` registers MP Health annotations as bean defining annotations -- except it doesn't define a default scope. We could just define a default scope of `@Singleton` for those additional BDAs and call it a day, no?

Ladicek replied:
I tried defining the default scope for the additional BDAs and removing the `annotationTransformer` method and all tests in the SmallRye Health extension pass. I claim that's good enough :-)

mkouba replied:
> Thinking about it again, I have no idea why SmallRyeHealthProcessor.annotationTransformer even exists. 

Most probably because the `BeanDefiningAnnotationBuildItem` originally did not have the `defaultScope` and seems to be implemented on top of stereotypes... which is a bit weird but works sooo.

Ladicek replied:
Yea I need to check why the SmallRye Health test failed in CI, it all passed locally.

mkouba commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:
```diff
@@ -281,6 +296,10 @@ public void transform(TransformationContext transformationContext) {
         }
+        // register additional stereotypes
+        for (StereotypeRegistrarBuildItem registrar : stereotypeRegistrars) {
+            builder.addStereotypeRegistrar(registrar.getStereotypeRegistrar());
```
Shouldn't we also deprecate `io.quarkus.arc.processor.BeanProcessor.Builder.setAdditionalStereotypes()`?

Ladicek replied:
This PR already does that: https://github.com/quarkusio/quarkus/pull/26264/files#diff-c9b2b8f7f82a42a3270b51243922c82ca8e6617e554a67788a60ac8381701391R545

Ladicek commented:
Rebased to get rid of the RestClient test failures.

mkouba commented:
The CI failure may be related?

mkouba commented:
> Yea I need to check why the SmallRye Health test failed in CI, it all passed locally.

So what's the status of this PR?

Ladicek commented:
I'm just getting back to this. (I wanted to work on #27574 and figured it would be best to have this done first).

Ladicek commented:
I can also just get rid of the SmallRye Health changes, as they are not really relevant to the subject of this PR. I just stumbled upon that piece of code and found it's suboptimal, so I figured I could just fix it and squeeze it in. I'll investigate for a bit and if I don't manage to find the issue, I'll get rid of the changes here.

Ladicek commented:
So the test failure from CI is:

```
Expected: iterable with items ["alpha1", "bravo1"] in any order
  Actual: <[alpha2, bravo1]>
```

The health checks are:

```java
static class HealthCheckProducers {
    static final AtomicInteger ALPHA_COUNTER = new AtomicInteger();
    static final AtomicInteger BRAVO_COUNTER = new AtomicInteger();

    // No scope - @Singleton is used by default
    @Readiness
    HealthCheck alpha() {
        int idx = ALPHA_COUNTER.incrementAndGet();
        return () -> HealthCheckResponse.builder().up().name("alpha" + idx).build();
    }

    @RequestScoped
    @Readiness
    HealthCheck bravo() {
        int idx = BRAVO_COUNTER.incrementAndGet();
        return () -> HealthCheckResponse.builder().up().name("bravo" + idx).build();
    }
}
```

So `alpha` has no scope annotation, thereby defaulting to singleton. There should only ever be `alpha1`. The existence of `alpha2` means the `alpha` health check has a different scope. I have no idea why.

Gonna revert those changes, then.

Ladicek committed:
Improve stereotypes support

This commit adds a better API for registering custom stereotypes, similar
to existing APIs: `StereotypeRegistrar` and `StereotypeRegistrarBuildItem`.
The existing `AdditionalStereotypeBuildItem` is very confusing to use, so
it is deprecated and all its usages are replaced with the new API.

Also, this commit allows defining stereotypes for synthetic beans. Scope,
name, alternative status and priority are all applied to the synthetic bean,
but interceptor bindings are not. We don't apply interceptors to synthetic
beans in general, this case is not an exception.

Ladicek commented:
Done.

mkouba commented:
> So `alpha` has no scope annotation, thereby defaulting to singleton. There should only ever be `alpha1`. The existence of `alpha2` means the `alpha` health check has a different scope. I have no idea why.

I believe that this would deserve a little investigation. Do you have a diff somewhere or a branch with a state before you reversed the change?

Ladicek commented:
I do: https://github.com/Ladicek/quarkus/commits/smallrye-health-default-scope The interesting thing is, the test didn't fail for me locally, and only failed in CI on Windows. I don't know what to take from it.

mkouba commented:
> The interesting thing is, the test didn't fail for me locally, and only failed in CI on Windows. I don't know what to take from it.

That's superweird.
